tools/syz-trace2syz/proggen/preprocess_hooks.go

```diff
@@ -0,0 +1,432 @@
+// Copyright 2018 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package proggen
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/prog"
+	"github.com/google/syzkaller/tools/syz-trace2syz/parser"
+)
+
+type pair struct {
+	A string
+	B string
+}
+type sock struct {
+	domain   uint64
+	level    uint64
+	protocol uint64
+}
+
+// CallVariantMap maps system calls to their variants (system calls with $ like socket$packet)
+// Keys represent the parts of the system call that need to be used
+// to identify the variant. Calls like socket require all three arguments
+// Some require two like setosckopt
+type CallVariantMap struct {
```

Comment:
I've spent a while reading this file. It seems to be doing 3 separate things:

1. The main one: choosing right discrimination for some syscalls.
2. Adding tailing 0 arguments to syscalls like open.
3. Doing some custom processing for shmget.

It seems that 2 can be removed entirely right away. We already use DefaultArg for missing args, so adding 0s looks unnecessary.
I think 3 should be moved elsewhere. It's a very custom thing for a single syscall. It does not feel like it should be intermixed with choosing syscall discriminations.
Now re 1. It has lots of duplication both across different syscalls (e.g. handling of const and flags arguments is duplicated in lots of places) and for each syscall (we duplicate discriminating args when we build a map and use it, this is a recipe for copy-paste errors, e.g. use arg 1 when building and arg 2 when selecting).
Also, combined handling of multiple syscalls (e.g. getsockopt and setsockopt) looks wrong. We can have a discrimination for setsockopt, but it does not mean we have the same for getsockopt. But since there is a single map for both, we can choose setsockopt prefix and append it to getsockopt and get a non-existent syscall name. The same applies to accept/bind/connect.

I think we need something like this for discrimination selection. This is data-driven and has absolute minimum of information per-syscall. The rest of the processing is generic. Code is not duplicated across syscalls and within syscall handling. Adding new syscalls is easy.
```go
var discriminatorArgs = map[string][]int{
	"bpf":         {0},
	"fcntl":       {1},
	"ioprio_get":  {0},
	"socket":      {0, 1, 2},
	"socketpair":  {0, 1, 2},
	"ioctl":       {0, 1},
	"getsockopt":  {1, 2},
	"setsockopt":  {1, 2},
	"accept":      {0},
	"accept4":     {0},
	"bind":        {0},
	"connect":     {0},
	"recvfrom":    {0},
	"sendto":      {0},
	"getsockname": {0},
}

type CallSelector struct {
	callCache map[string][]*prog.Syscall
}

// Select returns the best matching descrimination for this syscall.
func (cs *CallSelector) Select(ctx *Context, call *parser.Syscall) *prog.Syscall {
	match := ctx.Target.SyscallMap[call.CallName]
	discriminators := discriminatorArgs[call.CallName]
	if len(discriminators) == 0 {
		return match
	}
	score := 0
	for _, meta := range cs.callSet(ctx, call.CallName) {
		if score1 := matchCall(ctx, meta, call, discriminators); score1 > score {
			match, score = meta, score1
		}
	}
	return match
}

// callSet returns all syscalls with the given name.
func (cs *CallSelector) callSet(ctx *Context, callName string) []*prog.Syscall {
	calls, ok := cs.callCache[callName]
	if ok {
		return calls
	}
	for _, call := range ctx.Target.Syscalls {
		if call.CallName == callName {
			calls = append(calls, call)
		}
	}
	cs.callCache[callName] = calls
	return calls
}

// matchCall returns match score between meta and call.
// Higher score means better match, -1 if they are not matching at all.
func matchCall(ctx *Context, meta *prog.Syscall, call *parser.Syscall, discriminators []int) int {
	score := 0
	for _, i := range discriminators {
		if i >= len(meta.Args) || i >= len(call.Args) {
			return -1
		}
		typ := meta.Args[i]
		arg := call.Args[i]
		switch t := typ.(type) {
		case *prog.ConstType:
			// Consts must match precisely.
			expr, ok := arg.(parser.Expression)
			if !ok || expr.Eval(ctx.Target) != t.Val {
				return -1
			}
			score += 10
		case *prog.FlagsType:
			// Flags may or may not match, but matched flags increase score.
			expr, ok := arg.(parser.Expression)
			if !ok {
				return -1
			}
			val := expr.Eval(ctx.Target)
			for _, v := range t.Vals {
				if v == val {
					score++
					break
				}
			}
		case *prog.ResourceType:
			// Resources must match one of subtypes,
			// the more precise match, the higher the score.
			retArg := ctx.ReturnCache.get(t, arg)
			if retArg == nil {
				return -1
			}
			matched := false
			for i, kind := range retArg.Type().(*prog.ResourceType).Desc.Kind {
				if kind == t.Desc.Name {
					score += i + 1
					matched = true
					break
				}
			}
			if !matched {
				return -1
			}
		}
	}
	return score
}
```

Reply:
This seems quite good! I will add this. I also wonder if something similar to this could be used for generating unions. Given an intermediate type it seems like you want to select the union variant that has the highest score.

Reply:
The less hardcoded syscalls/unions we have, the better. Custom handling of every syscall and union is very hard to get right and it will contain off-by-one errors and get out of sync with descriptions.
Note that we can also e.g. rearrange union fields so that we select the first option more frequently.